src/app/browser/qgsinbuiltdataitemproviders.cpp

```diff
@@ -688,3 +691,133 @@ bool QgsProjectItemGuiProvider::handleDoubleClick( QgsDataItem *item, QgsDataIte
     return false;
   }
 }
+
+QString QgsFieldsItemGuiProvider::name()
+{
+  return QStringLiteral( "fields_item" );
+}
+
+void QgsFieldsItemGuiProvider::populateContextMenu( QgsDataItem *item, QMenu *menu, const QList<QgsDataItem *> &selectedItems, QgsDataItemGuiContext context )
+{
+  Q_UNUSED( selectedItems )
+
+  if ( !item || item->type() != QgsDataItem::Type::Fields )
+    return;
+
+
+  if ( QgsFieldsItem *fieldsItem = qobject_cast<QgsFieldsItem *>( item ) )
+  {
+    QgsProviderMetadata *md { QgsProviderRegistry::instance()->providerMetadata( fieldsItem->providerKey() ) };
+    if ( md )
+    {
+      std::unique_ptr<QgsAbstractDatabaseProviderConnection> conn { static_cast<QgsAbstractDatabaseProviderConnection *>( md->createConnection( fieldsItem->connectionUri(), {} ) ) };
+      // Check if it is supported
+      if ( conn && conn->capabilities().testFlag( QgsAbstractDatabaseProviderConnection::Capability::AddField ) )
+      {
+        QAction *addColumnAction = new QAction( tr( "New field…" ), menu );
+        connect( addColumnAction, &QAction::triggered, this, [ = ]
+        {
+          std::unique_ptr<QgsVectorLayer> layer { fieldsItem->layer( ) };
```

Comment:
Warning: it's not thread safe to access the item here. You need to collect all the useful information (uri as string, etc) when creating the action, and capture these only. There's a chance that a background thread may cause the item to be removed in between showing the menu and the user clicking an action (this isn't theoretical -- it's caused crashes in the past)

Reply:
Ok, any idea  about how to handle refresh ? https://github.com/qgis/QGIS/pull/37760/files#diff-0ed2a5629e6b6a189c65bd14659bafe4R730

Reply:
What if 

`connect( addColumnAction, &QAction::triggered, item, `
instead of
`connect( addColumnAction, &QAction::triggered, this, ` ?

it should disconnect the lambda if `item` is deleted.

Reply:
or maybe better pass a `QPointer(item)` to the lambda?

Reply:
Setting the connection scope looks like a good clean approach, I'd go with that one!